lib/autoupdate/agent.go

```diff
@@ -65,30 +86,204 @@ type AgentUpdateStatus struct {
+	// VersionsDir for installing Teleport (usually /var/lib/teleport/versions).
+	VersionsDir string
+}
+
+// AgentUpdater implements the agent-local logic for Teleport agent auto-updates.
 type AgentUpdater struct {
+	// Log contains a logger.
 	Log *slog.Logger
+	// Pool used for requests to the Teleport web API.
+	Pool *x509.CertPool
+	// InsecureSkipVerify skips TLS verification.
+	InsecureSkipVerify bool
+	// ConfigPath contains the path to the agent auto-updates configuration.
+	ConfigPath string
+	// Installer manages installations of the Teleport agent.
+	Installer AgentInstaller
+}
+
+// AgentInstaller provides an API for installing Teleport agents.
+type AgentInstaller interface {
+	// Install the Teleport agent at version from the download template.
+	Install(ctx context.Context, version, template string) error
+	// Remove the Teleport agent at version.
+	Remove(ctx context.Context, version string) error
+}
+
+// AgentUserConfig contains user overrides for installing Teleport agents.
+type AgentUserConfig struct {
+	// Proxy address, scheme and port optional.
+	// Overrides existing value if specified.
+	Proxy string
+	// Group identifier for updates (e.g., staging)
+	// Overrides existing value if specified.
+	Group string
+	// URLTemplate for the Teleport tgz download URL
+	// Overrides existing value if specified.
+	URLTemplate string
+	// ForceVersion to the specified version.
+	ForceVersion string
+}
+
+// Enable enables agent updates and attempts an initial update.
+// If the initial update succeeds, auto-updates are enabled and the configuration is persisted.
+// Otherwise, the auto-updates configuration is not changed.
+// This function is idempotent.
+func (u *AgentUpdater) Enable(ctx context.Context, userCfg AgentUserConfig) error {
+	// Read configuration from updates.yaml and override any new values passed as flags.
+	cfg, err := u.readConfig(u.ConfigPath)
+	if err != nil {
+		return trace.Errorf("failed to read update.yaml: %w", err)
+	}
+	if userCfg.Proxy != "" {
+		cfg.Spec.Proxy = userCfg.Proxy
+	}
+	if userCfg.Group != "" {
+		cfg.Spec.Group = userCfg.Group
+	}
+	if userCfg.URLTemplate != "" {
+		cfg.Spec.URLTemplate = userCfg.URLTemplate
+	}
+	cfg.Spec.Enabled = true
+	if err := validateUpdatesSpec(&cfg.Spec); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Lookup target version from the proxy.
+	addr, err := libutils.ParseAddr(cfg.Spec.Proxy)
+	if err != nil {
+		return trace.Errorf("failed to parse proxy server address: %w", err)
+	}
+
+	desiredVersion := userCfg.ForceVersion
+	if desiredVersion == "" {
+		resp, err := webclient.Find(&webclient.Config{
+			Context:   ctx,
+			ProxyAddr: addr.Addr,
+			Insecure:  u.InsecureSkipVerify,
+			Timeout:   30 * time.Second,
+			//Group:     cfg.Spec.Group, // TODO(sclevine): add web API
+			Pool: u.Pool,
+		})
+		if err != nil {
+			return trace.Errorf("failed to request version from proxy: %w", err)
+		}
+		desiredVersion, _ = "16.3.0", resp // TODO(sclevine): add web API
```

Comment:
nit: I would rather remove webapi request until its not in the find handler

Reply:
I'd like to keep it in if you don't mind, so it's easy to test the connection logic against real clusters.